pkg/descheduler/evictions/evictions.go

```diff
@@ -77,7 +255,144 @@ func NewPodEvictor(
 		metricsEnabled:             options.metricsEnabled,
 		nodePodCount:               make(nodePodEvictedCount),
 		namespacePodCount:          make(namespacePodEvictCount),
+		featureGates:               featureGates,
+	}
+
+	if featureGates.Enabled(features.EvictionsInBackground) {
+		erCache := newEvictionRequestsCache(assumedEvictionRequestTimeoutSeconds)
+
+		handlerRegistration, err := podInformer.AddEventHandler(
+			cache.ResourceEventHandlerFuncs{
+				AddFunc: func(obj interface{}) {
+					pod, ok := obj.(*v1.Pod)
+					if !ok {
+						klog.Error(nil, "Cannot convert to *v1.Pod", "obj", obj)
+						return
+					}
+					if _, exists := pod.Annotations[EvictionRequestAnnotationKey]; exists {
+						if _, exists := pod.Annotations[EvictionInProgressAnnotationKey]; exists {
+							// Ignore completed/suceeeded or failed pods
+							if pod.Status.Phase != v1.PodSucceeded && pod.Status.Phase != v1.PodFailed {
+								klog.V(3).Infof("Eviction in background detected. Adding pod %q to the cache.", klog.KObj(pod))
+								if err := erCache.addPod(pod); err != nil {
+									klog.ErrorS(err, "Unable to add pod to cache", "pod", pod)
+								}
+							}
+						}
+					}
+				},
+				UpdateFunc: func(oldObj, newObj interface{}) {
+					oldPod, ok := oldObj.(*v1.Pod)
+					if !ok {
+						klog.Error(nil, "Cannot convert oldObj to *v1.Pod", "oldObj", oldObj)
+						return
+					}
+					newPod, ok := newObj.(*v1.Pod)
+					if !ok {
+						klog.Error(nil, "Cannot convert newObj to *v1.Pod", "newObj", newObj)
+						return
+					}
+					if newPod.Annotations == nil {
+						return
+					}
+					// Ignore pod's that are not subject to an eviction in background
+					if _, exists := newPod.Annotations[EvictionRequestAnnotationKey]; !exists {
+						return
+					}
+					// Remove completed/suceeeded or failed pods from the cache
+					if newPod.Status.Phase == v1.PodSucceeded || newPod.Status.Phase == v1.PodFailed {
+						klog.V(3).Infof("Pod with eviction in background completed. Removing pod %q from the cache.", klog.KObj(newPod))
+						if err := erCache.deletePod(newPod); err != nil {
+							// If the deletion fails the cache may block eviction
+							klog.ErrorS(err, "Unable to delete updated pod from cache", "pod", newPod)
+						}
+						return
+					}
+					// Ignore any pod that does not have eviction in progress
+					if _, exists := newPod.Annotations[EvictionInProgressAnnotationKey]; !exists {
```

Comment:
Oh. I misunderstood. So it can happen a pod might still have an eviction in background correctly finished. Yet, if `EvictionRequestAnnotationKey` is removed (e.g. by accident or on purpose) the pod is no longer considered for further processing. `EvictionRequestAnnotationKey` is the referential one. We might still check for `EvictionInProgressAnnotationKey` annotation as well just in case there's one. Yet, some implementations might decide to just remove `EvictionRequestAnnotationKey` annotation to signal an eviction in background was aborted/canceled.

Reply:
> Yet, if EvictionRequestAnnotationKey is removed (e.g. by accident or on purpose) the pod is no longer considered for further processing.

I am not sure if I see all the loose ends here :D

If the pod was in the cache when the `Request` annotation was removed, then we skip evicting it (thus we process the `InProgress` annotation)

Reply:
There's always some space for errors. What's important is to properly remove entry from the cache eventually so all these "loose ends" are not persistent in the cache forever.

Reply:
> What's important is to properly remove entry from the cache
which we would not do when `evictionAssumed == false` and `EvictionRequestAnnotationKey` is removed

Reply:
Good catch. Updating the delete event handler to delete every pod that's in the cache: https://github.com/kubernetes-sigs/descheduler/pull/1466/commits/bb084d6e0f50ae99eafa4567e5f97dd31ab09c38

Reply:
That is much better. I still think it is unfortunate that we do not deschedule pods that had both `EvictionRequestAnnotationKey` and `EvictionInProgressAnnotationKey` removed at the same time. If the owner cancels the progress in this way and the pod is not terminating, then the pod will be immune to descheduling.

Reply:
Removing the pod from the cache during update event just in case the pod is in the cache: https://github.com/kubernetes-sigs/descheduler/pull/1466/commits/4425072b93da0b0c43c1fe37e9beec4e12b6ceb5

Reply:
looks good, thx!